src/lib/mdns/ServicePool.cpp

```diff
@@ -0,0 +1,272 @@
+#include "ServicePool.h"
+
+#include <string.h>
+
+#include <lib/support/CHIPMem.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/RandUtils.h>
+
+using chip::Mdns::kMdnsTextMaxSize;
+using chip::Mdns::MdnsService;
+using chip::Mdns::TextEntry;
+
+namespace {
+
+void FreeTextEntries(MdnsService * service)
+{
+    if (service->mTextEntryies)
+    {
+        for (size_t i = 0; i < service->mTextEntrySize; i++)
+        {
+            TextEntry & entry = service->mTextEntryies[i];
+
+            if (entry.mKey)
+            {
+                chip::Platform::MemoryFree(const_cast<char *>(entry.mKey));
+                entry.mKey = nullptr;
+            }
+            if (entry.mData)
+            {
+                chip::Platform::MemoryFree(const_cast<uint8_t *>(entry.mData));
+                entry.mData = nullptr;
+            }
+        }
+        chip::Platform::MemoryFree(service->mTextEntryies);
+        service->mTextEntryies = nullptr;
+    }
+}
+
+CHIP_ERROR CopyTextEntries(MdnsService * dest, const MdnsService & source)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+
+    if (source.mTextEntrySize)
+    {
+        dest->mTextEntryies = static_cast<TextEntry *>(chip::Platform::MemoryCalloc(source.mTextEntrySize, sizeof(TextEntry)));
+        VerifyOrExit(dest->mTextEntryies != nullptr, error = CHIP_ERROR_NO_MEMORY);
+        dest->mTextEntrySize = source.mTextEntrySize;
+
+        for (size_t i = 0; i < source.mTextEntrySize; i++)
+        {
+            TextEntry & entry = dest->mTextEntryies[i];
+            size_t keySize    = strnlen(source.mTextEntryies[i].mKey, kMdnsTextMaxSize);
+
+            VerifyOrExit(keySize < kMdnsTextMaxSize, error = CHIP_ERROR_INVALID_ARGUMENT);
+            entry.mKey = static_cast<char *>(chip::Platform::MemoryAlloc(keySize + 1));
+            VerifyOrExit(entry.mKey != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            strncpy(const_cast<char *>(entry.mKey), source.mTextEntryies[i].mKey, keySize + 1);
+            entry.mData = static_cast<uint8_t *>(chip::Platform::MemoryAlloc(source.mTextEntryies[i].mDataSize));
+            VerifyOrExit(entry.mData != nullptr, error = CHIP_ERROR_NO_MEMORY);
+            entry.mDataSize = source.mTextEntryies[i].mDataSize;
+            memcpy(const_cast<uint8_t *>(entry.mData), source.mTextEntryies[i].mData, entry.mDataSize);
+        }
+    }
+
+exit:
+    if (error != CHIP_NO_ERROR)
+    {
+        FreeTextEntries(dest);
+    }
+
+    return error;
+}
+
+} // namespace
+
+namespace chip {
+namespace Mdns {
+
+ServicePool::Entry::Entry() : mHasValue(false), mHasPending(false), mNodeId(0), mFabricId(0)
+{
+    mService.mTextEntryies  = nullptr;
+    mService.mTextEntrySize = 0;
+}
+
+ServicePool::Entry::Entry(Entry && rhs)
+{
+    MoveFrom(std::move(rhs));
+}
+
+ServicePool::Entry & ServicePool::Entry::operator=(Entry && rhs)
+{
+    return MoveFrom(std::move(rhs));
```

Comment:
Should the existing heap pointers we have be cleared/freed here?  Or can those not exist because this is only used in some cases when we are known to not have any?  Seems a little fragile to me....  See more on this below in the rehashing code.

Reply:
This move assignment is designed to take over all the existing heap pointers and mainly used for rehashing. In such circumstance the new entries will reuse the heap memory before the rehash.

Reply:
Removed.